### ISSUE TITLE
Implementing the Gateway interface with ACLs for synchronizing to CVMFS

### ISSUE THREAD

**Issue by mboisson** [ComputeCanada/software-stack#35]
Being able to have ACLs to allow only specific groups to update specific paths in CVMFS would possibly enable us to broaden the group of people who can install software. For example, if we can lock down more critical paths (such as the Lmod configuration, Gentoo/Nix, and the common profile scripts), we could allow more people to install software through EasyBuild.

**Comment by mboisson** [comment]
@rptaylor pointed to this information : https://cvmfs.readthedocs.io/en/stable/cpt-repository-gateway.html#advanced-repository-configuration

**Comment by mboisson** [comment]
From @rptaylor's comment in Slack : 
```
The API keys are easier to associate with a system than an individual user though. I don't think you can set up one publishing system with multiple keys.
But you could come up with e.g. a containerized setup where each person provides their own key and gets their own private publishing container, which would be useful in its own right anyway. I think there has been some work on that and there is some future development planned
```

**Comment by mboisson** [comment]
Jakob, from CVMFS development mentioned in a CVMFS-EESSI meeting that the Gateway interface is not yet battle-tested in production.

**Comment by mboisson** [comment]
This is a tentative plan for how to do this transition. 

- [ ] Switch from syncing through the build-node to switching through a dedicated `publisher-node`. This requires
  - [ ] Modify the stratum0 to accept sshfs and cvmfs commands from our `publisher-node` rather than `build-node`
  - [ ] Modify `rsnt-sync` to sync from a different path to dev (since CephFS is mounted under a different path)
- [ ] Switch `dev` repository to use the gateway interface. This requires:
  - [ ] Adjust our `start`, `publish`, `abort` scripts to use the gateway commands instead of the `ssh` commands for `dev`
  - [ ] Adjust `rsnt-sync` to sync `dev` to `/cvmfs/soft-dev.computecanada.ca` rather than through sshfs
  - [ ] Adjust `rsnt-sync` to sync prod from `/cvmfs/soft-dev.computecanada.ca` through to the sshfs location of prod
  - [ ] Modify the `dev` stratum0 to the gateway interface
- [ ] Switch `prod` repository to use the gateway interface. This requires:
  - [ ] Adjust our `start`, `publish`, `abort` scripts to use the gateway commands instead of the `ssh` commands for `prod`
  - [ ] Adjust `rsnt-sync` to sync `prod` to `/cvmfs/soft.computecanada.ca` from `/cvmfs/soft-dev.computecanada.ca` rather than through sshfs
  - [ ] Modify the `prod` stratum0 to the gateway interface
- [ ] Switch `restricted` repository to use the gateway interface. This requires:
  - [ ] Adjust our `start`, `publish`, `abort` scripts to use the gateway commands instead of the `ssh` commands for `restricted`
  - [ ] Adjust `rsnt-sync` to sync `restricted` to `/cvmfs/restricted.computecanada.ca` from `/cvmfs_ceph`
  - [ ] Modify the `restricted` stratum0 to the gateway interface
- [ ] Switch `data` repository to use the gateway interface. This requires:
  - [ ] Adjust our `start`, `publish`, `abort` scripts to use the gateway commands instead of the `ssh` commands for `data`
  - [ ] Adjust `rsnt-sync` to sync `restricted` to `/cvmfs/data.rsnt.computecanada.ca` from `/cvmfs_ceph`
  - [ ] Modify the `data.rsnt` stratum0 to the gateway interface

**Comment by rptaylor** [comment]
Here is some information about the new client-like publishing approach the developers are working on, alluded to in an earlier comment.
https://cvmfs.readthedocs.io/en/stable/cpt-enter.html

**Comment by mboisson** [comment]
We hit two issues while doing phase 3 and 4. 

First, when I first setup the production repository, there was a symlink in place from `/cvmfs/soft.computecanada.ca` to `/cvmfs_ceph/soft.computecanada.ca`, and `cvmfs_server`’s `mkfs` command started `chown`ing all files to `libuser` under the symlink. By the time I realized it and killed the mkfs operation, there were 151k files that had been changed to `libuser` as owner.
Other than `/cvmfs/soft.computecanada.ca/containers`, they were all under `easybuild`, so I chown’ed them all back to `ebuser`
The list of affected files is in
``` 
[root@publisher-node ~]# ls /root/libuser_owned_files*
/root/libuser_owned_files_easybuild.txt  /root/libuser_owned_files_second_pass.txt  /root/libuser_owned_files.txt
``` 
An issue has been created in CVMFS's Jira for this: 
https://sft.its.cern.ch/jira/browse/CVM-2068

Second, and this seems to be a quirk of the gateway publishing interface, it looks like permissions of the root directory are changing. This does not seem to impact the clients (confirmed by temporarily configuring a client to access the new stratum0 directly). 
https://cernvm-forum.cern.ch/t/repo-permissions-changing-after-abort/163

This however affects the visibility of the repositories on the `publisher-node` by users other than `libuser`, and basically means that other users can not read them. This should not be an issue however, as `publisher-node` should only be used to publish content.
